app/controllers/users/omniauth_callbacks_controller.rb

```diff
@@ -83,6 +80,16 @@ def complete
     redirect_to @origin
   end
 
+  def valid_origin?(uri)
```

Comment:
I think this method might be easier to read if we do it  without the if statement? Something like

```
return false if !uri 
return false if !(uri.host == nil || uri.host == Discourse.current_hostname)
return false if uri.path.start_with?.....
....
true
```